extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -0,0 +1,2040 @@
+                stageId.getQueryId(),
+                stageId.getStageNumber(),
+                IntStream.range(0, partitions.size())
+                         .mapToObj(i -> StringUtils.format("%s:%s", i, partitions.get(i)))
+                         .collect(Collectors.joining(", "))
+            );
+          } else {
+            log.info(
+                "Query [%s] sending out partition boundaries for stage %d.",
+                stageId.getQueryId(),
+                stageId.getStageNumber()
+            );
+          }
+
+          postResultPartitionBoundariesForStage(
+              queryDef,
+              stageId.getStageNumber(),
+              queryKernel.getResultPartitionBoundariesForStage(stageId),
+              queryKernel.getWorkerInputsForStage(stageId).workers()
+          );
+        }
+      }
+
+      logKernelStatus(queryDef.getQueryId(), queryKernel);
+
+      // Live reports: update stage phases, worker counts, partition counts.
+      for (StageId stageId : queryKernel.getActiveStages()) {
+        final int stageNumber = stageId.getStageNumber();
+        stagePhasesForLiveReports.put(stageNumber, queryKernel.getStagePhase(stageId));
+
+        if (queryKernel.doesStageHaveResultPartitions(stageId)) {
+          stagePartitionCountsForLiveReports.computeIfAbsent(
+              stageNumber,
+              k -> Iterators.size(queryKernel.getResultPartitionsForStage(stageId).iterator())
+          );
+        }
+
+        stageWorkerCountsForLiveReports.putIfAbsent(
+            stageNumber,
+            queryKernel.getWorkerInputsForStage(stageId).workerCount()
+        );
+      }
+
+      // Live reports: update stage end times for any stages that just ended.
+      for (StageId stageId : queryKernel.getActiveStages()) {
+        if (ControllerStagePhase.isSuccessfulTerminalPhase(queryKernel.getStagePhase(stageId))) {
+          stageRuntimesForLiveReports.compute(
+              queryKernel.getStageDefinition(stageId).getStageNumber(),
+              (k, currentValue) -> {
+                if (currentValue.getEnd().equals(DateTimes.MAX)) {
+                  return new Interval(currentValue.getStart(), DateTimes.nowUtc());
+                } else {
+                  return currentValue;
+                }
+              }
+          );
+        }
+      }
+
+      // Notify the workers to clean up the stages which can be marked as finished.
+      cleanUpEffectivelyFinishedStages(queryDef, queryKernel);
+
+      if (!queryKernel.isDone()) {
+        // Run the next command, waiting for it if necessary.
+        Consumer<ControllerQueryKernel> command = kernelManipulationQueue.take();
+        command.accept(queryKernel);
+
+        // Run all pending commands after that one. Helps avoid deep queues.
+        // After draining the command queue, move on to the next iteration of the controller loop.
+        while ((command = kernelManipulationQueue.poll()) != null) {
```

Comment:
Looks like commands are run synchronous with stage management. Presumably the stage management stuff could get stuck if there is a network error, which means we'd not get to process commands. Are the command functionally determined by having live workers? That there will be no commands unless all workers are up and we get this far? Or, are there race conditions to consider? There is a delay in starting worker 2, and during that time, worker 1 sends requests which are ignored until worker 2 finally comes up?

I guess the question is, rather than this being a big synchronous loop, should the above be an async state machine? Moving between states concurrent with processing messages from the user (to cancel, say) or workers that are up?

Since status updates go into the queue, a stalled worker start could end up creating a large backlog of worker updates, and so the user will not see the actual state of the task during such stalls.

Async state machines are difficult, so there is no expectation for this PR to be changed. This is an suggestion for the future direction.

Reply:
> Presumably the stage management stuff could get stuck if there is a network error, which means we'd not get to process commands.

Yes, that's a risk. I think it'd be better to structure the code such that RPCs to workers happen asynchronously. That structure didn't make it into this particular version of the code. However I think it would be an important future change.

> are there race conditions to consider? There is a delay in starting worker 2, and during that time, worker 1 sends requests which are ignored until worker 2 finally comes up?

In this scenario, the commands originating from worker 1 would go into the queue and be handled after worker 2 becomes responsive. It's not ideal, because it creates the possibility for slow response to worker 1 and, in extreme cases, for the queue to get full of messages. It'd be better to send out asynchronous RPCs instead of synchronous ones. I had imagined this would be done in a future patch.